File: harness/graph_snapshot_contract/snapshot.py
"""Structured-set seals and final GraphSnapshot identity/record construction."""

from __future__ import annotations

from governance_contract import ContractError

from .codec import domain_digest, self_digest
from .constants import (
    CANONICALIZATION, COVERAGE_DOMAIN, CROSSWALK_SET_DOMAIN, EDGE_SET_DOMAIN,
    EXTRACTOR_SET_DOMAIN, IDENTITY_FIELDS, MAX_EDGE_UNION, MAX_SNAPSHOT_BYTES,
    NODE_SET_DOMAIN, PROFILE_ID, RESULT, SNAPSHOT_API, SNAPSHOT_DOMAIN,
    SNAPSHOT_IDENTITY_DOMAIN, SOURCE_SET_DOMAIN, SYSTEM_UNKNOWN_REASONS,
    UNRESOLVED_EDGE_SET_DOMAIN, UNRESOLVED_NODE_SET_DOMAIN,
)
from .coverage import build_coverage, build_freshness
from .records import ensure_global_identity_uniqueness, structured_set_digest


def _set_digests(sources, extractors, nodes, edges, unresolved_nodes,
                 unresolved_edges, crosswalk):
    return {
        "source_set_sha256": structured_set_digest(SOURCE_SET_DOMAIN, sources),
        "extractor_set_sha256": structured_set_digest(EXTRACTOR_SET_DOMAIN, extractors),
        "node_set_sha256": structured_set_digest(NODE_SET_DOMAIN, nodes),
        "edge_set_sha256": structured_set_digest(EDGE_SET_DOMAIN, edges),
        "unresolved_node_set_sha256": structured_set_digest(
            UNRESOLVED_NODE_SET_DOMAIN, unresolved_nodes),
        "unresolved_edge_set_sha256": structured_set_digest(
            UNRESOLVED_EDGE_SET_DOMAIN, unresolved_edges),
        "crosswalk_set_sha256": structured_set_digest(
            CROSSWALK_SET_DOMAIN, crosswalk),
    }


def build_snapshot(graph: dict[str, object], project_id: str, request_sha256: str,
                   sources, extractors, nodes, edges, unresolved_nodes,
                   unresolved_edges, crosswalk) -> dict[str, object]:
    if len(edges) > MAX_EDGE_UNION:
        raise ContractError("ADR-0065 resolved edge union limit exceeded")
    coverage = build_coverage(graph, len(nodes), len(edges))
    digests = _set_digests(sources, extractors, nodes, edges, unresolved_nodes,
                           unresolved_edges, crosswalk)
    coverage_sha = domain_digest(
        COVERAGE_DOMAIN, coverage, max_bytes=MAX_SNAPSHOT_BYTES)
    snapshot = {
        "adr_0062_node_crosswalk": crosswalk, "api_version": SNAPSHOT_API,
        "canonicalization": CANONICALIZATION, "coverage": coverage,
        "coverage_sha256": coverage_sha, "edges": edges,
        "extractors": extractors, "freshness": build_freshness(graph),
        "nodes": nodes, "profile_id": PROFILE_ID, "project_id": project_id,
        "request_sha256": request_sha256, "result": RESULT,
        "snapshot_id": "", "snapshot_identity_sha256": "",
        "snapshot_sha256": "", "sources": sources,
        "system_knowledge_status": "unknown",
        "system_unknown_reason_codes": list(SYSTEM_UNKNOWN_REASONS),
        "unresolved_edges": unresolved_edges, "unresolved_nodes": unresolved_nodes,
        **digests,
    }
    identity = {field: snapshot[field] for field in IDENTITY_FIELDS["snapshot"]}
    identity_sha = domain_digest(
        SNAPSHOT_IDENTITY_DOMAIN, identity, max_bytes=MAX_SNAPSHOT_BYTES)
    snapshot["snapshot_identity_sha256"] = identity_sha
    snapshot["snapshot_id"] = "graph-snapshot-" + identity_sha
    ensure_global_identity_uniqueness([
        (sources, "source_identity_sha256"),
        (extractors, "extractor_identity_sha256"),
        (nodes, "node_identity_sha256"),
        (edges, "edge_identity_sha256"),
        (unresolved_nodes, "unresolved_node_identity_sha256"),
        (unresolved_edges, "unresolved_edge_identity_sha256"),
        (crosswalk, "adr_0062_node_sha256"),
    ], identity_sha)
    snapshot["snapshot_sha256"] = self_digest(
        SNAPSHOT_DOMAIN, snapshot, "snapshot_sha256", max_bytes=MAX_SNAPSHOT_BYTES)
    return snapshot

File: forge-core/internal/graphsnapshot/snapshot.go
package graphsnapshot

import "fmt"

func (value *projector) buildSnapshot(requestSHA256 string) (Snapshot, []byte, error) {
	nodes, crosswalk, err := value.buildNodes()
	if err != nil {
		return Snapshot{}, nil, err
	}
	edges, err := value.buildEdges()
	if err != nil {
		return Snapshot{}, nil, err
	}
	unresolvedNodes, err := value.buildUnresolvedNodes()
	if err != nil {
		return Snapshot{}, nil, err
	}
	unresolvedEdges, err := value.buildUnresolvedEdges()
	if err != nil {
		return Snapshot{}, nil, err
	}
	if err := value.validateProjectionCounts(nodes, edges, unresolvedNodes, unresolvedEdges, crosswalk); err != nil {
		return Snapshot{}, nil, err
	}
	coverage := value.buildCoverage(nodes, edges)
	snapshot := Snapshot{
		ADR0062NodeCrosswalk: crosswalk, APIVersion: snapshotVersion,
		Canonicalization: canonicalization, Coverage: coverage, Edges: edges,
		Extractors: []Extractor{value.extractor}, Freshness: Freshness{
			ExpiresAtUnixMS:  value.observation.ObservedAtUnixMS,
			ObservedAtUnixMS: value.observation.ObservedAtUnixMS,
			ReasonCodes:      append([]string{}, freshnessReasons...), Status: "unknown",
		}, Nodes: nodes, ProfileID: value.profile.profileID, ProjectID: value.projectID,
		RequestSHA256: requestSHA256, Result: value.profile.resultText,
		Sources: []Source{value.source}, SystemKnowledgeStatus: "unknown",
		SystemUnknownReasonCodes: append([]string{}, value.profile.systemUnknownReasons...),
		UnresolvedEdges:          unresolvedEdges, UnresolvedNodes: unresolvedNodes,
	}
	if err := sealSnapshotSets(&snapshot); err != nil {
		return Snapshot{}, nil, err
	}
	if err := value.sealSnapshotIdentity(&snapshot); err != nil {
		return Snapshot{}, nil, err
	}
	encoded, err := canonicalJSON(snapshot, maxSnapshotBytes)
	return snapshot, encoded, err
}

func (value *projector) validateProjectionCounts(
	nodes []Node,
	edges []Edge,
	unresolvedNodes []UnresolvedNode,
	unresolvedEdges []UnresolvedEdge,
	crosswalk []Crosswalk,
) error {
	packageCount := len(value.observation.Packages)
	testCount := 0
	if value.profile.includeTestSources {
		testCount = len(value.testNodesByKey)
	}
	if len(nodes) != packageCount+testCount+1 || len(crosswalk) != packageCount ||
		len(unresolvedNodes) != len(value.observation.Diagnostics)+len(value.observation.Module.NestedModules) {
		return fmt.Errorf("node or crosswalk projection is not bijective")
	}
	localCount := 0
	for _, dependency := range value.observation.Dependencies {
		if dependency.Resolution == "local" {
			localCount++
		}
	}
	if len(edges) != packageCount+testCount+localCount ||
		localCount+len(unresolvedEdges) != len(value.observation.Dependencies) {
		return fmt.Errorf("dependency projection is not bijective")
	}
	return validateAggregateLocators(
		nodes, edges, unresolvedNodes, unresolvedEdges, value.profile.maxAggregateLocators)
}

func validateAggregateLocators(
	nodes []Node,
	edges []Edge,
	unresolvedNodes []UnresolvedNode,
	unresolvedEdges []UnresolvedEdge,
	limit int,
) error {
	total := 0
	for _, item := range nodes {
		total += len(item.SourceLocators)
	}
	for _, item := range edges {
		total += len(item.SourceLocators)
	}
	for _, item := range unresolvedNodes {
		total += len(item.SourceLocators)
	}
	for _, item := range unresolvedEdges {
		total += len(item.SourceLocators)
	}
	if total > limit {
		return fmt.Errorf("aggregate source locators exceed %d", limit)
	}
	return nil
}

func sealSnapshotSets(value *Snapshot) error {
	var err error
	if value.SourceSetSHA256, err = setDigest(sourceSetDomain, value.Sources); err != nil {
		return err
	}
	if value.ExtractorSetSHA256, err = setDigest(extractorSetDomain, value.Extractors); err != nil {
		return err
	}
	if value.NodeSetSHA256, err = setDigest(nodeSetDomain, value.Nodes); err != nil {
		return err
	}
	if value.EdgeSetSHA256, err = setDigest(edgeSetDomain, value.Edges); err != nil {
		return err
	}
	return sealRemainingSets(value)
}

func sealRemainingSets(value *Snapshot) error {
	var err error
	if value.UnresolvedNodeSetSHA256, err = setDigest(unresolvedNodeSetDomain, value.UnresolvedNodes); err != nil {
		return err
	}
	if value.UnresolvedEdgeSetSHA256, err = setDigest(unresolvedEdgeSetDomain, value.UnresolvedEdges); err != nil {
		return err
	}
	if value.CrosswalkSetSHA256, err = setDigest(crosswalkSetDomain, value.ADR0062NodeCrosswalk); err != nil {
		return err
	}
	value.CoverageSHA256, err = digestValue(coverageDomain, value.Coverage)
	return err
}

func (worker *projector) sealSnapshotIdentity(value *Snapshot) error {
	identity := snapshotIdentity{
		CoverageSHA256: value.CoverageSHA256, CrosswalkSetSHA256: value.CrosswalkSetSHA256,
		EdgeSetSHA256: value.EdgeSetSHA256, ExtractorSetSHA256: value.ExtractorSetSHA256,
		NodeSetSHA256: value.NodeSetSHA256, ProfileID: value.ProfileID,
		ProjectID: value.ProjectID, RequestSHA256: value.RequestSHA256,
		SourceSetSHA256:         value.SourceSetSHA256,
		UnresolvedEdgeSetSHA256: value.UnresolvedEdgeSetSHA256,
		UnresolvedNodeSetSHA256: value.UnresolvedNodeSetSHA256,
	}
	digest, err := digestValue(snapshotIdentityDomain, identity)
	if err != nil {
		return err
	}
	value.SnapshotIdentitySHA256 = digest
	value.SnapshotID = "graph-snapshot-" + digest
	if err := worker.claimIdentity("snapshot", digest, value.SnapshotID); err != nil {
		return err
	}
	value.SnapshotSHA256 = ""
	value.SnapshotSHA256, err = digestValue(snapshotDomain, *value)
	return err
}
